wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,346 @@
+        return not self.__eq__(other)
+
+    def first(self) -> Optional[Program]:
+        if self.pair:
+            return self.pair[0]
+        return None
+
+    def rest(self) -> Optional[Program]:
+        if self.pair:
+            return self.pair[1]
+        return None
+
+    def listp(self) -> bool:
+        return self.pair is not None
+
+    def nullp(self) -> bool:
+        return self.atom == b""
+
+    def list_len(self) -> int:
+        c = 0
+        v: CLVMStorage = self
+        while v.pair is not None:
+            v = v.pair[1]
+            c += 1
+        return c
+
+    def at(self, position: str) -> Optional["Program"]:
+        """
+        Take a string of `f` and `r` characters and follow that path.
+
+        Example:
+
+        ```
+        p1 = Program.to([10, 20, 30, [15, 17], 40, 50])
+        assert Program.to(17) == at(p1, "rrrfrf")
+        ```
+
+        Returns `None` if an atom is hit at some intermediate node.
+
+        ```
+        p1 = Program.to(10)
+        assert None == at(p1, "rr")
+        ```
+        """
+        r = at(self, position)
+        if r is None:
+            return r
+        return self.to(r)
+
+    def at_many(self, *positions: str) -> List[Optional["Program"]]:
+        """
+        Call `.at` multiple times.
+
+        Why? So you can write
+
+        `
+        if p.at_many("f", "rf", "rfrf") == [5, 10, 15]:
+        `
+        instead of
+        `
+        if [p.at("f"), p.at("rf"), p.at("rfrf")] == [5, 10, 15]:
+        `
+        """
+        return [self.at(_) for _ in positions]
+
+    def replace(self, **kwargs) -> "Program":
+        """
+        Create a new program replacing the given paths (using `at` syntax).
+        Example:
+        ```
+        >>> p1 = Program.to([100, 200, 300])
+        >>> print(p1.replace(f=105) == Program.to([105, 200, 300]))
+        True
+        >>> p2 = [100, 200, [301, 302]]
+        >>> print(p1.replace(rrf=[301, 302]) == Program.to(p2))
+        True
+        >>> p2 = [105, 200, [301, 302]]
+        >>> print(p1.replace(f=105, rrf=[301, 302]) == Program.to(p2))
+        True
+        ```
+
+        This is a convenience method intended for use in the wallet or
+        command-line hacks where it would be easier to morph elements
+        of an existing clvm object tree than to rebuild one from scratch.
+
+        Note that `Program` objects are immutable. This function returns a
+        new object; the original is left as-is.
+        """
+        return self.to(replace(self, **kwargs))
+
+    def tree_hash(self) -> bytes:
+        # we operate on the unwrapped version to prevent the re-wrapping that
+        # happens on each invocation of `Program.pair` whenever possible
+        if self._cached_sha256_treehash is None:
+            self._cached_sha256_treehash = sha256_treehash(self._unwrapped)
+        return self._cached_sha256_treehash
+
+    def run_with_cost(
+        self, args, max_cost: int = MAX_COST, flags: int = 0
```

Comment:
this is a bit unrelated, but wondering how come we chose to default to MAX_COST and not BLOCK_COST? Assuming that most people will write clvm programs to run on chain not as a scripting language.

Reply:
This is the classic problem of how when you go down a stack of calls, the parameters and call names go from more specific to more generic. IMO, this level should know as little as possible about consensus, and encoding a consensus constant at this level might temp users to depend on it when they really probably shouldn't.

Reply:
I think it's a security risk to have this `MAX_COST` as a default parameter. I can't think of any situation where we *want* to use this as the cost limit

Reply:
Point taken. I need to think more about how to handle this. Perhaps a `Program` class-level `DEFAULT_MAX_COST` that can be set by applications at start-up time. I'm a little hesitant to set policy on this that knows about Chia consensus though.

Reply:
I did this with a `UNSAFE_MAX_COST` value, which must be set prior to using `.run`.